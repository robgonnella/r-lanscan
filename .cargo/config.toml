[target.'cfg(any(target_os = "macos", target_os = "linux"))']
runner = 'sudo -E'
