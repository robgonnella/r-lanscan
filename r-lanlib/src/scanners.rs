use std::collections::HashSet;

use serde;
use serde::{Deserialize, Serialize};

pub const IDLE_TIMEOUT: u16 = 10000;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum PortStatus {
    Closed,
    Open,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Port {
    pub id: u16,
    pub service: String,
    pub status: PortStatus,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum DeviceStatus {
    Offline,
    Online,
}

// ARP Result from a single device
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Device {
    pub hostname: String,
    pub ip: String,
    pub mac: String,
    pub status: DeviceStatus,
    pub vendor: String,
}

// Device with open ports
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceWithPorts {
    pub ip: String,
    pub mac: String,
    pub hostname: String,
    pub vendor: String,
    pub open_ports: HashSet<Port>,
}

// SYN Result from a single device
#[derive(Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct SYNScanResult {
    pub device: Device,
    pub open_port: Port,
}

#[derive(Debug)]
pub enum ScanMessage {
    Done(()),
    ARPScanResult(Device),
    SYNScanResult(SYNScanResult),
}

impl ScanMessage {
    pub fn is_arp_message(&self) -> Option<&Device> {
        match self {
            ScanMessage::Done(_msg) => None,
            ScanMessage::SYNScanResult(_msg) => None,
            ScanMessage::ARPScanResult(msg) => Some(msg),
        }
    }

    pub fn is_syn_message(&self) -> Option<&SYNScanResult> {
        match self {
            ScanMessage::Done(_msg) => None,
            ScanMessage::SYNScanResult(msg) => Some(msg),
            ScanMessage::ARPScanResult(_msg) => None,
        }
    }

    pub fn is_done(&self) -> Option<()> {
        match self {
            ScanMessage::Done(_msg) => Some(()),
            ScanMessage::SYNScanResult(_msg) => None,
            ScanMessage::ARPScanResult(_msg) => None,
        }
    }
}

pub trait Scanner: Sync + Send {
    fn scan(&self);
}

pub mod arp_scanner;
pub mod full_scanner;
pub mod syn_scanner;
