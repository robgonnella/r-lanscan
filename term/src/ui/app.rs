use color_eyre::eyre::{Context, Result};
use core::time;
use log::*;
use ratatui::{
    backend::TestBackend,
    crossterm::{
        event::{
            self, DisableMouseCapture, EnableMouseCapture, Event as CrossTermEvent, KeyCode,
            KeyModifiers,
        },
        execute,
        terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
    },
    layout::Rect,
    prelude::CrosstermBackend,
    Terminal,
};
use std::{
    cell::RefCell,
    io::{self, Stdout},
    sync::{
        mpsc::{Receiver, Sender},
        Arc,
    },
};

use super::{
    events::types::Event,
    store::{action::Action, Store},
    views::{
        main::MainView,
        traits::{CustomWidgetContext, View},
    },
};

type Backend = CrosstermBackend<Stdout>;

pub struct App {
    terminal: RefCell<Terminal<Backend>>,
    // here to enable unit tests - not an ideal solution but okay for now
    test_terminal: Option<Terminal<TestBackend>>,
    store: Arc<Store>,
    main_view: Box<dyn View>,
    event_loop_sender: Sender<Event>,
    event_loop_receiver: Receiver<Event>,
}

pub fn create_app(tx: Sender<Event>, rx: Receiver<Event>, store: Arc<Store>) -> Result<App> {
    // setup terminal
    enable_raw_mode().wrap_err("failed to enter raw mode")?;
    let mut stdout = io::stdout();
    execute!(stdout, EnterAlternateScreen, EnableMouseCapture)
        .wrap_err("failed to enter alternate screen")?;
    let backend = CrosstermBackend::new(stdout);
    let terminal = Terminal::new(backend).wrap_err("failed to create terminal")?;
    Ok(App::new(tx, rx, terminal, store))
}

impl App {
    fn new(
        tx: Sender<Event>,
        rx: Receiver<Event>,
        terminal: Terminal<Backend>,
        store: Arc<Store>,
    ) -> Self {
        Self {
            terminal: RefCell::new(terminal),
            test_terminal: None,
            store: Arc::clone(&store),
            main_view: Box::new(MainView::new(store, tx.clone())),
            event_loop_sender: tx,
            event_loop_receiver: rx,
        }
    }

    // only exposed in tests to enable unit testing App
    // not an ideal solution but okay for now
    #[cfg(test)]
    fn new_test(
        tx: Sender<Event>,
        rx: Receiver<Event>,
        terminal: Terminal<Backend>,
        test_terminal: Terminal<TestBackend>,
        store: Arc<Store>,
    ) -> Self {
        Self {
            terminal: RefCell::new(terminal),
            test_terminal: Some(test_terminal),
            store: Arc::clone(&store),
            main_view: Box::new(MainView::new(store, tx.clone())),
            event_loop_sender: tx,
            event_loop_receiver: rx,
        }
    }

    pub fn launch(&self) -> Result<()> {
        self.start_app_loop()?;
        self.exit()?;
        Ok(())
    }

    fn start_app_loop(&self) -> Result<()> {
        loop {
            let state = self.store.get_state();

            if state.ui_paused {
                if let Ok(evt) = self.event_loop_receiver.recv()
                    && evt == Event::ResumeUI {
                        self.restart()?;
                        self.store.dispatch(Action::SetUIPaused(false));
                        self.event_loop_sender.send(Event::UIResumed)?;
                        continue;
                    }
            } else if let Ok(evt) = self.event_loop_receiver.try_recv()
                && evt == Event::PauseUI {
                    self.pause()?;
                    self.store.dispatch(Action::SetUIPaused(true));
                    self.event_loop_sender.send(Event::UIPaused)?;
                    continue;
                }

            let mut ctx = CustomWidgetContext {
                state: state.clone(),
                app_area: Rect::default(),
                events: self.event_loop_sender.clone(),
            };

            if self.test_terminal.is_some() {
                // app is under test - just draw once and exit
                // not an ideal solution but okay for now
                let mut terminal = self.test_terminal.clone().unwrap();
                let _ = terminal.draw(|f| {
                    ctx = CustomWidgetContext {
                        state,
                        app_area: f.area(),
                        events: self.event_loop_sender.clone(),
                    };
                    self.main_view.render_ref(f.area(), f.buffer_mut(), &ctx)
                });
                return Ok(());
            }

            self.terminal.borrow_mut().draw(|f| {
                ctx = CustomWidgetContext {
                    state,
                    app_area: f.area(),
                    events: self.event_loop_sender.clone(),
                };
                self.main_view.render_ref(f.area(), f.buffer_mut(), &ctx)
            })?;

            // Use poll here so we don't block the thread, this will allow
            // rendering of incoming device data from network as it's received
            if let Ok(has_event) = event::poll(time::Duration::from_millis(60))
                && has_event {
                    let evt = event::read()?;

                    let handled = self.main_view.process_event(&evt, &ctx);

                    if let CrossTermEvent::Key(key) = evt {
                        match key.code {
                            KeyCode::Char('q') => {
                                // allow overriding q key
                                if !handled {
                                    self.event_loop_sender.send(Event::Quit)?;
                                    return Ok(());
                                }
                            }
                            KeyCode::Char('c') => {
                                // do not allow overriding ctrl-c
                                if key.modifiers == KeyModifiers::CONTROL {
                                    info!("APP RECEIVED CONTROL-C SEQUENCE");
                                    self.event_loop_sender.send(Event::Quit)?;
                                    return Ok(());
                                }
                            }
                            _ => {}
                        }
                    }
                }
        }
    }

    fn pause(&self) -> Result<()> {
        self.exit()?;
        self.store.dispatch(Action::SetUIPaused(true));
        Ok(())
    }

    fn restart(&self) -> Result<()> {
        if self.test_terminal.is_none() {
            let mut terminal = self.terminal.borrow_mut();
            enable_raw_mode()?;
            execute!(
                terminal.backend_mut(),
                EnterAlternateScreen,
                EnableMouseCapture
            )?;
            terminal.hide_cursor()?;
            terminal.clear()?;
            self.store.dispatch(Action::SetUIPaused(false));
        }
        Ok(())
    }

    fn exit(&self) -> Result<()> {
        if self.test_terminal.is_none() {
            let mut terminal = self.terminal.borrow_mut();
            disable_raw_mode()?;
            execute!(
                terminal.backend_mut(),
                LeaveAlternateScreen,
                DisableMouseCapture
            )?;
            terminal.show_cursor()?;
        }
        Ok(())
    }
}

#[cfg(test)]
#[path = "./app_tests.rs"]
mod tests;
