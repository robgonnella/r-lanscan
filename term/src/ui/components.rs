pub mod field;
pub mod footer;
pub mod header;
pub mod input;
pub mod popover;
pub mod scrollbar;
pub mod table;
