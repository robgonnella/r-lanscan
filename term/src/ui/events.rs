mod commander;
pub mod manager;
pub mod types;
