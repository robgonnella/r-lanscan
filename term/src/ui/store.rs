use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
};

use crate::{
    config::{ConfigManager, DEFAULT_CONFIG_ID},
    ui::colors::Theme,
};

pub mod action;
pub mod derived;
pub mod reducer;
pub mod state;

/**
 * Manages the state of our application
 */
pub struct Store {
    state: Mutex<state::State>,
    reducer: reducer::Reducer,
}

impl Store {
    pub fn new(config_manager: Arc<Mutex<ConfigManager>>) -> Self {
        let config = config_manager
            .lock()
            .unwrap()
            .get_by_id(DEFAULT_CONFIG_ID)
            .unwrap();

        let true_color_enabled = match supports_color::on(supports_color::Stream::Stdout) {
            Some(support) => support.has_16m,
            _ => false,
        };

        let theme = Theme::from_string(&config.theme);
        let colors = crate::ui::colors::Colors::new(
            theme.to_palette(true_color_enabled),
            true_color_enabled,
        );

        Self {
            reducer: reducer::Reducer::new(config_manager),
            state: Mutex::new(state::State {
                true_color_enabled,
                ui_paused: false,
                error: None,
                render_view_select: false,
                view_id: state::ViewID::Devices,
                config,
                arp_history: HashMap::new(),
                devices: Vec::new(),
                device_map: HashMap::new(),
                selected_device: None,
                selected_device_config: None,
                colors,
                message: None,
                cmd_in_progress: None,
                cmd_output: None,
            }),
        }
    }

    pub fn dispatch(&self, action: action::Action) {
        let mut prev_state = self.state.lock().unwrap();
        let new_state = self.reducer.reduce(prev_state.clone(), action);
        *prev_state = new_state;
    }

    pub fn get_state(&self) -> state::State {
        self.state.lock().unwrap().clone()
    }
}

#[cfg(test)]
#[path = "./store_tests.rs"]
mod tests;
