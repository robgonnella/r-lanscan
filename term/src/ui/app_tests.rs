use nanoid::nanoid;
use ratatui::backend::TestBackend;
use std::{
    fs,
    sync::{Mutex, mpsc},
};

use crate::config::ConfigManager;

use super::*;

fn setup() -> (String, Arc<Store>, App) {
    fs::create_dir_all("generated").unwrap();
    let tmp_path = format!("generated/{}.yml", nanoid!());
    let conf_manager = Arc::new(Mutex::new(ConfigManager::new(tmp_path.as_str())));
    let store = Arc::new(Store::new(conf_manager));
    let (tx, rx) = mpsc::channel();
    let stdout = io::stdout();
    let real_terminal = Terminal::new(CrosstermBackend::new(stdout)).unwrap();
    let test_terminal = Terminal::new(TestBackend::new(80, 40)).unwrap();
    let app = App::new_test(tx, rx, real_terminal, test_terminal, Arc::clone(&store));
    (tmp_path, store, app)
}

fn tear_down(conf_path: String) {
    fs::remove_file(conf_path).unwrap();
}

#[test]
fn test_app() {
    let (conf_path, _store, app) = setup();
    let _ = app.launch();
    tear_down(conf_path);
}
