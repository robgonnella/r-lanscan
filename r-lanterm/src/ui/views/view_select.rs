use std::{cell::RefCell, sync::Arc};

use itertools::Itertools;
use ratatui::crossterm::event::{Event, KeyCode, KeyEventKind};

use crate::ui::{
    components::table::{self, Table},
    store::{action::Action, state::ViewID, store::Store},
};

use super::traits::{CustomWidgetContext, CustomWidgetRef, EventHandler, View};

pub struct ViewSelect {
    store: Arc<Store>,
    view_ids: Vec<ViewID>,
    table: RefCell<Table>,
}

impl ViewSelect {
    pub fn new(view_ids: Vec<ViewID>, padding: usize, store: Arc<Store>) -> Self {
        let mut spacer = String::from("");

        if padding > 0 {
            for _ in 0..padding {
                spacer += " ";
            }
        }

        let table_items = view_ids
            .clone()
            .iter()
            .map(|v| vec![format!("{}{}", spacer, v.to_string())])
            .collect_vec();

        let mut table_select = Table::new(
            table_items,
            None,
            vec![15; view_ids.len()],
            table::DEFAULT_ITEM_HEIGHT,
        );

        table_select.next();

        Self {
            store,
            view_ids,
            table: RefCell::new(table_select),
        }
    }

    fn next(&self) {
        self.table.borrow_mut().next();
    }

    fn previous(&self) {
        self.table.borrow_mut().previous();
    }

    fn handle_selected(&self) {
        let i = self.table.borrow().selected();
        if let Some(selected) = i {
            let id = self.view_ids[selected].clone();
            self.store.dispatch(Action::UpdateView(id));
            self.store.dispatch(Action::ToggleViewSelect);
        }
    }
}

impl View for ViewSelect {
    fn id(&self) -> ViewID {
        ViewID::ViewSelect
    }
}

impl EventHandler for ViewSelect {
    fn process_event(&self, evt: &Event, ctx: &CustomWidgetContext) -> bool {
        if !ctx.state.render_view_select {
            return false;
        }

        let mut handled = false;

        match evt {
            Event::Key(key) => {
                if key.kind == KeyEventKind::Press {
                    match key.code {
                        KeyCode::Char('j') | KeyCode::Down => {
                            self.next();
                            handled = true;
                        }
                        KeyCode::Char('k') | KeyCode::Up => {
                            self.previous();
                            handled = true;
                        }
                        KeyCode::Esc => {
                            if ctx.state.render_view_select {
                                self.store.dispatch(Action::ToggleViewSelect);
                                handled = true;
                            }
                        }
                        KeyCode::Enter => {
                            self.handle_selected();
                            handled = true;
                        }
                        _ => {}
                    }
                }
            }
            _ => {}
        }

        handled
    }
}

impl CustomWidgetRef for ViewSelect {
    fn render_ref(
        &self,
        area: ratatui::prelude::Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        self.table.borrow().render_ref(area, buf, ctx);
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use nanoid::nanoid;
    use pnet::util::MacAddr;
    use r_lanlib::scanners::{DeviceWithPorts, Port};
    use ratatui::{backend::TestBackend, Terminal};
    use std::{
        collections::{HashMap, HashSet},
        fs,
        sync::Mutex,
    };

    use crate::config::{Config, ConfigManager};

    use super::*;

    fn setup() -> (ViewSelect, Arc<Store>, String) {
        fs::create_dir_all("generated").unwrap();
        let tmp_path = format!("generated/{}.yml", nanoid!());
        let conf_manager = Arc::new(Mutex::new(ConfigManager::new(tmp_path.as_str())));
        let store = Arc::new(Store::new(conf_manager));
        let config = Config {
            id: "default".to_string(),
            cidr: "192.168.1.1/24".to_string(),
            default_ssh_identity: "id_rsa".to_string(),
            default_ssh_port: "22".to_string(),
            default_ssh_user: "user".to_string(),
            device_configs: HashMap::new(),
            ports: vec![],
            theme: "Blue".to_string(),
        };
        store.dispatch(Action::CreateAndSetConfig(config));

        let mut open_ports: HashSet<Port> = HashSet::new();
        open_ports.insert(Port {
            id: 80,
            service: "http".to_string(),
        });

        let device_1 = DeviceWithPorts {
            hostname: "hostname".to_string(),
            ip: "10.10.10.1".to_string(),
            is_current_host: false,
            mac: MacAddr::default().to_string(),
            open_ports: open_ports.clone(),
            vendor: "mac".to_string(),
        };

        let device_2 = DeviceWithPorts {
            hostname: "dev2_hostname".to_string(),
            ip: "10.10.10.2".to_string(),
            is_current_host: true,
            mac: "ff:ff:ff:ff:ff:ff".to_string(),
            open_ports,
            vendor: "linux".to_string(),
        };

        store.dispatch(Action::AddDevice(device_1.clone()));
        store.dispatch(Action::AddDevice(device_2.clone()));
        let view_ids = vec![ViewID::Devices, ViewID::Config];
        (
            ViewSelect::new(view_ids, 2, Arc::clone(&store)),
            store,
            tmp_path,
        )
    }

    fn tear_down(conf_path: String) {
        fs::remove_file(conf_path).unwrap();
    }

    #[test]
    fn test_view_select_view() {
        let (view_select, store, conf_path) = setup();
        let mut terminal = Terminal::new(TestBackend::new(80, 15)).unwrap();
        let state = store.get_state();
        let channel = std::sync::mpsc::channel();

        terminal
            .draw(|frame| {
                let ctx = CustomWidgetContext {
                    state,
                    app_area: frame.area(),
                    events: channel.0,
                };

                view_select.render_ref(frame.area(), frame.buffer_mut(), &ctx);
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
        tear_down(conf_path);
    }
}
