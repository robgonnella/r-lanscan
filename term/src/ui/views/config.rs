use crate::ui::{
    colors::Theme,
    components::{
        field::Field,
        header::Header,
        input::{Input, InputState},
    },
    store::{
        action::Action,
        state::{State, ViewID},
        Store,
    },
};
use itertools::Itertools;
use ratatui::{
    crossterm::event::{Event, KeyCode, KeyEventKind},
    layout::{Constraint, Layout, Rect},
    widgets::Widget,
};
use std::{cell::RefCell, sync::Arc};

use super::traits::{
    CustomStatefulWidget, CustomWidget, CustomWidgetContext, CustomWidgetRef, EventHandler, View,
};

const THEMES: [Theme; 4] = [Theme::Blue, Theme::Emerald, Theme::Indigo, Theme::Red];

#[derive(Debug, Clone)]
enum Focus {
    SSHUser,
    SSHPort,
    SSHIdentity,
    Theme,
    ScanPorts,
}

pub struct ConfigView {
    store: Arc<Store>,
    theme_index: RefCell<usize>,
    editing: RefCell<bool>,
    focus: RefCell<Focus>,
    ssh_user_state: RefCell<InputState>,
    ssh_port_state: RefCell<InputState>,
    ssh_identity_state: RefCell<InputState>,
    theme_state: RefCell<InputState>,
    scan_ports_state: RefCell<InputState>,
}

impl ConfigView {
    pub fn new(store: Arc<Store>) -> Self {
        let state = store.get_state();

        let theme = Theme::from_string(&state.config.theme);

        let (idx, _) = THEMES
            .iter()
            .enumerate()
            .find(|(_, t)| **t == theme)
            .unwrap();

        Self {
            store,
            theme_index: RefCell::new(idx),
            editing: RefCell::new(false),
            focus: RefCell::new(Focus::SSHUser),
            ssh_user_state: RefCell::new(InputState {
                editing: false,
                value: String::from(""),
            }),
            ssh_port_state: RefCell::new(InputState {
                editing: false,
                value: String::from(""),
            }),
            ssh_identity_state: RefCell::new(InputState {
                editing: false,
                value: String::from(""),
            }),
            theme_state: RefCell::new(InputState {
                editing: false,
                value: String::from(""),
            }),
            scan_ports_state: RefCell::new(InputState {
                editing: false,
                value: String::from(""),
            }),
        }
    }

    fn next_color(&self) {
        let new_idx = *self.theme_index.borrow() + 1;
        *self.theme_index.borrow_mut() = new_idx % THEMES.len();
        let theme = THEMES[*self.theme_index.borrow()].clone();
        self.theme_state.borrow_mut().value = theme.to_string();
        self.store.dispatch(Action::PreviewTheme(theme));
    }

    fn previous_color(&self) {
        let count = THEMES.len();
        let new_idx = *self.theme_index.borrow() + count - 1;
        *self.theme_index.borrow_mut() = new_idx % count;
        let theme = THEMES[*self.theme_index.borrow()].clone();
        self.theme_state.borrow_mut().value = theme.to_string();
        self.store.dispatch(Action::PreviewTheme(theme));
    }

    fn set_config(&self, state: &State) {
        let mut config = state.config.clone();
        config.theme = THEMES[*self.theme_index.borrow()].clone().to_string();
        config.default_ssh_user = self.ssh_user_state.borrow().value.clone();
        let mut port = self.ssh_port_state.borrow().value.clone().parse::<u16>();
        if port.is_err() {
            port = Ok(22)
        }
        config.default_ssh_port = port.unwrap().to_string();
        config.default_ssh_identity = self.ssh_identity_state.borrow().value.clone();
        config.ports = self
            .scan_ports_state
            .borrow()
            .value
            .clone()
            .split(",")
            .map_into()
            .collect();
        self.store.dispatch(Action::UpdateConfig(config));
    }

    fn render_label(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let header = Header::new("Config".to_string());
        header.render(area, buf, ctx);
    }

    fn render_network(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let field = Field::new("Network".to_string(), ctx.state.config.cidr.clone());
        field.render(area, buf);
    }

    fn push_input_char(&self, char: char) {
        match *self.focus.borrow() {
            Focus::SSHUser => self.ssh_user_state.borrow_mut().value.push(char),
            Focus::SSHPort => self.ssh_port_state.borrow_mut().value.push(char),
            Focus::SSHIdentity => self.ssh_identity_state.borrow_mut().value.push(char),
            Focus::ScanPorts => self.scan_ports_state.borrow_mut().value.push(char),
            _ => {}
        };
    }

    fn pop_input_char(&self) {
        match *self.focus.borrow() {
            Focus::SSHUser => {
                self.ssh_user_state.borrow_mut().value.pop();
            }
            Focus::SSHPort => {
                self.ssh_port_state.borrow_mut().value.pop();
            }
            Focus::SSHIdentity => {
                self.ssh_identity_state.borrow_mut().value.pop();
            }
            Focus::ScanPorts => {
                self.scan_ports_state.borrow_mut().value.pop();
            }
            _ => {}
        };
    }

    fn reset_input_state(&self) {
        self.ssh_user_state.borrow_mut().editing = false;
        self.ssh_port_state.borrow_mut().editing = false;
        self.ssh_identity_state.borrow_mut().editing = false;
        self.theme_state.borrow_mut().editing = false;
        self.scan_ports_state.borrow_mut().editing = false;
    }

    fn focus_next(&self) {
        let next_focus = match *self.focus.borrow() {
            Focus::SSHUser => {
                if *self.editing.borrow() {
                    self.ssh_user_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = true;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = false;
                    self.scan_ports_state.borrow_mut().editing = false;
                }
                Focus::SSHPort
            }
            Focus::SSHPort => {
                if *self.editing.borrow() {
                    self.ssh_user_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = true;
                    self.theme_state.borrow_mut().editing = false;
                    self.scan_ports_state.borrow_mut().editing = false;
                }
                Focus::SSHIdentity
            }
            Focus::SSHIdentity => {
                if *self.editing.borrow() {
                    self.ssh_user_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = true;
                    self.scan_ports_state.borrow_mut().editing = false;
                }
                Focus::Theme
            }
            Focus::Theme => {
                if *self.editing.borrow() {
                    self.ssh_user_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = false;
                    self.scan_ports_state.borrow_mut().editing = true;
                }
                Focus::ScanPorts
            }
            Focus::ScanPorts => {
                if *self.editing.borrow() {
                    self.ssh_user_state.borrow_mut().editing = true;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = false;
                    self.scan_ports_state.borrow_mut().editing = false;
                }
                Focus::SSHUser
            }
        };

        *self.focus.borrow_mut() = next_focus;
    }

    fn focus_previous(&self) {
        let next_focus = match *self.focus.borrow() {
            Focus::ScanPorts => {
                if *self.editing.borrow() {
                    self.scan_ports_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = true;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_user_state.borrow_mut().editing = false;
                }
                Focus::Theme
            }
            Focus::Theme => {
                if *self.editing.borrow() {
                    self.scan_ports_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = true;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_user_state.borrow_mut().editing = false;
                }
                Focus::SSHIdentity
            }
            Focus::SSHIdentity => {
                if *self.editing.borrow() {
                    self.scan_ports_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = true;
                    self.ssh_user_state.borrow_mut().editing = false;
                }
                Focus::SSHPort
            }
            Focus::SSHPort => {
                if *self.editing.borrow() {
                    self.scan_ports_state.borrow_mut().editing = false;
                    self.theme_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_user_state.borrow_mut().editing = true;
                }
                Focus::SSHUser
            }
            Focus::SSHUser => {
                if *self.editing.borrow() {
                    self.scan_ports_state.borrow_mut().editing = true;
                    self.theme_state.borrow_mut().editing = false;
                    self.ssh_identity_state.borrow_mut().editing = false;
                    self.ssh_port_state.borrow_mut().editing = false;
                    self.ssh_user_state.borrow_mut().editing = false;
                }
                Focus::SSHIdentity
            }
        };

        *self.focus.borrow_mut() = next_focus;
    }

    fn render_ports(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let port_list = Input::new("Scanning Ports");
        port_list.render(area, buf, &mut self.scan_ports_state.borrow_mut(), ctx);
    }

    fn render_ssh(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        state: &State,
        ctx: &CustomWidgetContext,
    ) {
        let rects = Layout::vertical([
            Constraint::Length(1),
            Constraint::Length(1), // spacer
            Constraint::Length(1),
            Constraint::Length(1), // spacer
            Constraint::Length(1),
        ])
        .split(area);

        if !*self.editing.borrow() {
            self.ssh_user_state.borrow_mut().value = state.config.default_ssh_user.clone();
            self.ssh_port_state.borrow_mut().value = state.config.default_ssh_port.clone();
            self.ssh_identity_state.borrow_mut().value = state.config.default_ssh_identity.clone();
            self.theme_state.borrow_mut().value = state.config.theme.clone();
            self.scan_ports_state.borrow_mut().value = state.config.ports.join(",");
        }

        let ssh_user = Input::new("Default SSH User");
        let ssh_port = Input::new("Default SSH Port");
        let ssh_ident = Input::new("Default SSH Identity");

        ssh_user.render(rects[0], buf, &mut self.ssh_user_state.borrow_mut(), ctx);
        ssh_port.render(rects[2], buf, &mut self.ssh_port_state.borrow_mut(), ctx);
        ssh_ident.render(
            rects[4],
            buf,
            &mut self.ssh_identity_state.borrow_mut(),
            ctx,
        );
    }

    fn render_theme(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let input = Input::new("Theme <->");
        input.render(area, buf, &mut self.theme_state.borrow_mut(), ctx);
    }
}

impl View for ConfigView {
    fn id(&self) -> ViewID {
        ViewID::Config
    }
    fn legend(&self, _state: &State) -> &str {
        if *self.editing.borrow() {
            "(esc) exit configuration | (tab) focus next | (enter) save config"
        } else {
            "(c) configure"
        }
    }
    fn override_main_legend(&self, _state: &State) -> bool {
        *self.editing.borrow()
    }
}

impl CustomWidgetRef for ConfigView {
    fn render_ref(
        &self,
        area: Rect,
        buf: &mut ratatui::prelude::Buffer,
        ctx: &CustomWidgetContext,
    ) {
        let view_rects = Layout::vertical([
            Constraint::Length(1), // label
            Constraint::Length(1), // spacer
            Constraint::Length(1), // network
            Constraint::Length(1), // spacer
            Constraint::Length(5), // ssh
            Constraint::Length(1), // spacer
            Constraint::Length(1), // theme
            Constraint::Length(1), // spacer
            Constraint::Min(1),    // ports
        ])
        .split(area);

        let label_rects = Layout::horizontal([Constraint::Length(20)]).split(view_rects[0]);

        self.render_label(label_rects[0], buf, ctx);
        self.render_network(view_rects[2], buf, ctx);
        self.render_ssh(view_rects[4], buf, &ctx.state, ctx);
        self.render_theme(view_rects[6], buf, ctx);
        self.render_ports(view_rects[8], buf, ctx);
    }
}

impl EventHandler for ConfigView {
    fn process_event(&self, evt: &Event, ctx: &CustomWidgetContext) -> bool {
        if ctx.state.render_view_select {
            return false;
        }

        let mut handled = false;
        match evt {
            Event::FocusGained => {}
            Event::FocusLost => {}
            Event::Mouse(_m) => {}
            Event::Paste(_s) => {}
            Event::Resize(_x, _y) => {}
            Event::Key(key) => {
                if key.kind == KeyEventKind::Press {
                    match key.code {
                        KeyCode::Esc => {
                            if *self.editing.borrow() {
                                self.reset_input_state();
                                *self.focus.borrow_mut() = Focus::SSHUser;
                                *self.editing.borrow_mut() = false;
                                handled = true;
                            }
                        }
                        KeyCode::Tab => {
                            if *self.editing.borrow() {
                                self.focus_next();
                                handled = true;
                            }
                        }
                        KeyCode::BackTab => {
                            if *self.editing.borrow() {
                                self.focus_previous();
                                handled = true;
                            }
                        }
                        KeyCode::Right => {
                            if *self.editing.borrow() && self.theme_state.borrow().editing {
                                self.next_color();
                                handled = true;
                            }
                        }
                        KeyCode::Left => {
                            if *self.editing.borrow() && self.theme_state.borrow().editing {
                                self.previous_color();
                                handled = true;
                            }
                        }
                        KeyCode::Enter => {
                            if *self.editing.borrow() {
                                self.set_config(&ctx.state);
                                self.reset_input_state();
                                *self.focus.borrow_mut() = Focus::SSHUser;
                                *self.editing.borrow_mut() = false;
                                handled = true;
                            }
                        }
                        KeyCode::Backspace => {
                            if *self.editing.borrow() && !self.theme_state.borrow().editing {
                                self.pop_input_char();
                                handled = true;
                            }
                        }
                        KeyCode::Char(c) => {
                            if *self.editing.borrow() && !self.theme_state.borrow().editing {
                                // handle value update for focused element
                                self.push_input_char(c);
                                handled = true;
                            } else if c == 'c' && !*self.editing.borrow() {
                                // enter edit mode
                                *self.editing.borrow_mut() = true;
                                self.ssh_user_state.borrow_mut().editing = true;
                                handled = true;
                            }
                        }
                        _ => {}
                    }
                }
            }
        }

        handled
    }
}

#[cfg(test)]
#[path = "./config_tests.rs"]
mod tests;
