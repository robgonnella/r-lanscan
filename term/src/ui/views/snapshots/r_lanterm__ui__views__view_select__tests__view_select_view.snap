---
source: r-lanterm/src/ui/views/view_select.rs
expression: terminal.backend()
---
"                                                                                "
"  Devices                                                                     █ "
"                                                                              █ "
"                                                                              █ "
"  Config                                                                      █ "
"                                                                              █ "
"                                                                              █ "
"                                                                              █ "
"                                                                              █ "
"                                                                              █ "
"                                                                              █ "
"                                                                              █ "
"                                                                              ║ "
"                                                                              ║ "
"                                                                                "
